plugin/core/protocol.py

```diff
@@ -108,12 +121,31 @@ class SignatureHelpTriggerKind:
 
 Location = TypedDict('Location', {
     'uri': DocumentUri,
-    'range': Dict[str, Any]
+    'range': Range_T
 }, total=True)
 
+DocumentSymbol = TypedDict('DocumentSymbol', {
+    'name': str,
+    'detail': Optional[str],
+    'kind': int,
+    'tags': Optional[List[int]],
+    'deprecated': Optional[bool],
+    'range': Range_T,
+    'selectionRange': Range_T,
+    'children': Optional[List[Any]]  # mypy doesn't support recurive types like Optional[List[DocumentSymbol]]
```

Comment:
Are you just talking about it not being able to see definitions that are not yet defined? In that case you can use a string - `Optional[List['DocumentSymbol']]`

Reply:
No mypy doesn't support recursive types:

```
JsonDict = Dict[str, Union[None, int, float, JsonList, JsonDict]
```
This doesn't work because JsonDict references itself again. https://github.com/python/mypy/issues/731